File: 383-RansomNote/383-RansomNote.js
// Last updated: 4/29/2026, 10:30:42 AM
/**
 * @param {string} ransomNote
 * @param {string} magazine
 * @return {boolean}
 */
var canConstruct = function(ransomNote, magazine) {
    const hash = {};
    for(const ch of magazine) {
        if(hash[ch] === undefined) {
            hash[ch] = 1;
        } else {
            hash[ch]++;   
        }
    }
    for(const ch of ransomNote) {
        if(!hash[ch]) {
            return false;
        } else {
            hash[ch]--;
        }
    }
    return true;
};

File: 383-RansomNote/383-RansomNote.py
# Last updated: 4/29/2026, 10:30:46 AM
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        hash = {}
        for ch in magazine:
            if hash.get(ch) == None:
                hash[ch] = 1
            else: 
                hash[ch] += 1
        print(hash)
        
        for ch in ransomNote:
            if hash.get(ch) != None and hash[ch] > 0:
                hash[ch] -= 1
            else:
                return False
        
        return True